### PR TITLE
RenERC20 rate to support continuous fees

### PR DESCRIPTION
**Pull request by noiach**
This PR updated RenERC20 tokens to track a rate which dictates the ratio of tokens to the native asset they represent. This allows for continuous fees to be charged by updating this rate, without modifying the ERC20 balances on-chain.

### CODE REVIEW

File: contracts/Gateway/Gateway.sol [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        // The recipient must not be empty. Better validation is possible,
+        // but would need to be customized for each destination ledger.
+        require(_to.length != 0, "Gateway: to address is empty");
+
+        // Burn full amount and mint fee
+        uint256 absoluteFee = _amount.mul(burnFee).div(BIPS_DENOMINATOR);
+        uint256 amountScaled = token.fromUnderlying(_amount);
+        uint256 receivedScaled = token.fromUnderlying(_amount.sub(absoluteFee));
+        uint256 receivedValue = token.toUnderlying(receivedScaled);
+        uint256 amountDifference = amountScaled.sub(receivedScaled);
```
Comment (noiach) [inline contracts/Gateway/Gateway.sol]: This fee calculation is a bit complicated to try and match the exact fee values before - I'll work on simplifying it though.

File: contracts/Gateway/ERC20WithRate.sol [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    function balanceOfUnderlying(address _account)
+        public
+        view
+        returns (uint256)
+    {
+        return toUnderlying(balanceOf(_account));
+    }
+
+    function toUnderlying(uint256 _value) public view returns (uint256) {
```
Comment (loongy) [inline contracts/Gateway/ERC20WithRate.sol]: For consistency, `_value` should be `_amount`.

File: contracts/Gateway/ERC20WithRate.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        return _value.mul(_rate).div(_rateScale);
+    }
+
+    function fromUnderlying(uint256 _value) public view returns (uint256) {
```
Comment (loongy) [inline contracts/Gateway/ERC20WithRate.sol]: For consistency, `_value` should be `_amountUnderlying`.

File: contracts/Gateway/Gateway.sol [hunk 4]
```diff
@@ -192,17 +193,30 @@ contract Gateway is IGateway, Claimable, CanReclaimTokens {
         }
         status[signedMessageHash] = true;
 
+        uint256 amountScaled = token.fromUnderlying(_amount);
```
Comment (loongy) [inline contracts/Gateway/Gateway.sol]: The `_amount` should be called `_amountUnderlying`.

File: contracts/Gateway/Gateway.sol [hunk 5]
```diff
@@ -192,17 +193,30 @@ contract Gateway is IGateway, Claimable, CanReclaimTokens {
...
+        uint256 absoluteFeeScaled = amountScaled.mul(mintFee).div(
+            BIPS_DENOMINATOR
+        );
+        uint256 receivedAmountScaled = amountScaled.sub(absoluteFeeScaled);
+        token.mint(msg.sender, receivedAmountScaled);
+        token.mint(feeRecipient, absoluteFeeScaled);
 
         // Emit a log with a unique identifier 'n'.
+        // Use underlying amount, not scaled by rate.
+        uint256 receivedAmount = token.toUnderlying(receivedAmountScaled);
```
Comment (loongy) [inline contracts/Gateway/Gateway.sol]: The `receivedAmount` should be called `receivedAmountUnderlying`.

File: contracts/Gateway/Gateway.sol [hunk 6]
```diff
@@ -215,30 +229,34 @@ contract Gateway is IGateway, Claimable, CanReclaimTokens {
     ///        Bitcoin address.
     /// @param _amount The amount of the token being burnt, in its
     ///        smallest value. (e.g. satoshis for BTC)
-    function burn(bytes memory _to, uint256 _amount) public returns (uint256) {
+    function burnUnderlying(bytes memory _to, uint256 _amount)
```
Comment (loongy) [inline contracts/Gateway/Gateway.sol]: We should remove this function and require everyone to use `burn`. This leaves less surface for people to use the wrong function.

File: contracts/Gateway/Gateway.sol [hunk 7]
```diff
@@ -215,30 +229,34 @@ contract Gateway is IGateway, Claimable, CanReclaimTokens {
...
         require(_to.length != 0, "Gateway: to address is empty");
 
         // Burn full amount and mint fee
         uint256 absoluteFee = _amount.mul(burnFee).div(BIPS_DENOMINATOR);
-        token.burn(msg.sender, _amount);
-        token.mint(feeRecipient, absoluteFee);
-
-        // Emit a log with a unique identifier 'n'.
-        uint256 receivedValue = _amount.sub(absoluteFee);
+        uint256 amountScaled = token.fromUnderlying(_amount);
```
Comment (loongy) [inline contracts/Gateway/Gateway.sol]: The `_amount` should be `_amountUnderlying`.

File: contracts/Gateway/RenERC20.sol [hunk 8]
```diff
@@ -1,20 +1,39 @@
...
+        string memory _symbol,
+        uint8 _decimals
+    ) public initializer {
+        ERC20Detailed.initialize(_name, _symbol, _decimals);
+        ERC20WithRate.initialize(_nextOwner, _initialRate);
+        Claimable.initialize(_nextOwner);
+        CanReclaimTokens.initialize(_nextOwner);
+    }
 
     function burn(address _from, uint256 _amount) public onlyOwner {
```
Comment (loongy) [inline contracts/Gateway/RenERC20.sol]: Both `burn` and `mint` on the `RenERC20` should have a comment reminding people that cross-chain minting/burning must be done via the `Gateway`.

**Pull request merged**